File: go/signing.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"math/rand"
	"strings"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// Replace with your own values
const (
	embedPath   = "your path here"
	clientID    = "your clientid here"
	embedSecret = "your secret here"
)

func generateJWTEmbedURl() string {
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"sub":          "xyz@xyz.com",
		"jti":          fmt.Sprintf("%x", rand.Int63()),
		"iat":          time.Now().Unix(),
		"exp":          time.Now().Add(time.Hour * 1).Unix(),
		"iss":          clientID,
		"ver":          "1.1",
		"aud":          "sigmacomputing",
		"teams":        [...]string{"EmbedTeam"},
		"account_type": "Pro",
	})

	tokenString, err := token.SignedString([]byte(embedSecret))

	if err != nil {
		panic(err)
	}

	url := fmt.Sprintf("https://app.sigmacomputing.com/<your org>/<your workbook>?:embed=true&:jwt=%s", tokenString)
	return url
}

func myQuote(val string) string {
	val = strings.ReplaceAll(val, " ", "%20")
	return val
}

func urlencode(pairs map[string]interface{}) string {
	var encodedParams []string
	for k, v := range pairs {
		encodedParams = append(encodedParams, myQuote(k)+"="+myQuote(fmt.Sprint(v)))
	}
	return strings.Join(encodedParams, "&")
}

func secureEmbedUrl() string {
	params := map[string]interface{}{
		":nonce":              fmt.Sprintf("%x", rand.Int63()),
		":email":              "xyz@xyz.com",
		":external_user_id":   "xyz@xyz.com",
		":client_id":          clientID,
		":time":               time.Now().Unix(),
		":session_length":     3600,
		":mode":               "userbacked",
		":external_user_team": "EmbedTeam",
		":account_type":       "embedUser",
		// custom controls/parameters
		// "Store-Region": "West",
	}

	urlWithParams := embedPath + "?" + urlencode(params)

	hmacHash := hmac.New(sha256.New, []byte(embedSecret))
	io.WriteString(hmacHash, urlWithParams)
	signature := hex.EncodeToString(hmacHash.Sum(nil))

	urlWithSignature := urlWithParams + "&" + urlencode(map[string]interface{}{":signature": signature})

	return urlWithSignature
}

func main() {
	fmt.Println("=========JWT Embed URL=========")
	fmt.Println(generateJWTEmbedURl())
	fmt.Println("==================================")
	fmt.Println("=========Secure Embed URL=========")
	fmt.Println(secureEmbedUrl())
	fmt.Println("==================================")
}

File: python/signing.py
import hashlib
import hmac
import time
import io

try:
    from urllib.parse import quote
except:
    from urllib import quote
import uuid

# Replace with your own values
EMBED_PATH   = 'your path here'
CLIENT_ID    = 'your clientid here'
EMBED_SECRET = 'your secret here'

def urlencode(pairs):
    # Sigma doesn't support percent-encoded colons and commas in date range params
    # Percent-encoding those characters isn't required by RFC, so explicitly don't
    # encode them!
    def myquote(val):
        return quote(
            str(val),
            safe=",:",
        )

    return "&".join(myquote(k) + "=" + myquote(v) for k, v in pairs.items())


params = {
    ":nonce": uuid.uuid4(),
    ":email": "xyz@xyz.com",
    ":external_user_id": "1",
    ":client_id": CLIENT_ID,
    ":time": time.time(),
    ":session_length": 3600,
    ":mode": "view",
    ":mode": "userbacked",
    ":external_user_team": "Embedded Users,EmbeddingTown",
    ":account_type": "embedUser",
    # custom controls/parameters
    # "Store-Region": "West",
}


url_with_params = EMBED_PATH + "?" + urlencode(params)
    
signature = hmac.new(EMBED_SECRET.encode('utf-8'), url_with_params.encode('utf-8'), hashlib.sha256).hexdigest()
    
url_with_signature = url_with_params + "&" + urlencode({":signature": signature})
    
print(url_with_signature)